File: src/Search/Dijkstras.php
<?php

namespace GulDmitry\Algorithms\Search;

/**
 * Directed Acyclic Graph with positive weights.
 */
class Dijkstras
{
    public function search(array $graph, array $costs, array $parents): array
    {
        $processed = [];

        $findLowestExcludeProcessed = function (array $costs, array $processed) {
            $filtered = array_filter($costs, function ($k) use ($costs, $processed) {
                return !in_array($k, $processed);
            }, ARRAY_FILTER_USE_KEY);
            return empty($filtered) ? null : array_keys($filtered, min($filtered))[0];
        };

        $node = $findLowestExcludeProcessed($costs, $processed);
        while ($node !== null) {
            $cost = $costs[$node];
            $neighbors = $graph[$node];

            foreach ($neighbors as $k => $v) {
                $newCost = $cost + $v;

                if ($newCost < $costs[$k]) {
                    $costs[$k] = $newCost;
                    $parents[$k] = $node;
                }
            }
            $processed[] = $node;
            $node = $findLowestExcludeProcessed($costs, $processed);
        }
        return $parents;
    }
}

File: src/Search/dijkstras.go
package Search

import (
	"math"
)

func Dijkstras(
	graph map[string]map[string]float64,
	costs map[string]float64,
	parents map[string]string,
) map[string]string {
	path := map[string]string{}
	for k, v := range parents {
		path[k] = v
	}
	processed := []string{}

	findLowestElement := func() string {
		result := ""
		lowestValue := math.Inf(0)
	OUTER:
		for k, v := range costs {
			for _, v := range processed {
				if k == v {
					continue OUTER
				}
			}
			if v < lowestValue {
				result = k
			}
		}
		return result
	}

	node := findLowestElement()
	for node != "" {
		cost := costs[node]
		neighbors := graph[node]

		for k, neighborCost := range neighbors {
			newCost := cost + neighborCost

			if newCost < costs[k] {
				costs[k] = newCost
				path[k] = node
			}
		}
		processed = append(processed, node)
		node = findLowestElement()
	}

	return path
}
